Convert remaining JLL stdlibs to LazyLibraries

IanButterworth opened the pull request:
Builds on https://github.com/JuliaLang/julia/pull/58405 (which I rebased, so the merge base here is master for now)

TODO:
- [x] Convert the new Zstd_jll from #58344 
- [x] ~Convert LLD_jll?~ No, it's an executable not a lib
- [x] ~Convert p7zip_jll?~ No, it's an executable not a lib
- [x] ~Blocked by --trim failure~ #57707 Instead it copies libzstd over for testing with a const path
- [x] Fix linux RR test failure - release https://github.com/JuliaWeb/OpenSSL.jl/pull/40

staticfloat committed:
Convert more stdlibs to LazyLibraries. Add tests for dep tree.

This converts more JLL stdlibs to use lazily-loaded libraries, and as a
useful side-effect, causes them to be loaded by absolute path, isolating
them all from `LD_LIBRARY_PATH`-like effects.

Adds tests that audits the expressed library dependencies in our
LazyLibrary JLL definitions, to ensure that we don't get out of sync
with the actual binaries.

Co-Authored-By: Mosè Giordano <765740+giordano@users.noreply.github.com>
Co-Authored-By: Elliot Saba <staticfloat@gmail.com>

topolarity commented:
Unfortunately the `--trim` [failure](https://buildkite.com/julialang/julia-master/builds/47645#0196dbae-4bc4-490d-83fa-9c56691552f9/821-876) is expected until we resolve https://github.com/JuliaLang/julia/issues/57707:
```julia
Core.MethodError(f=Base.Libc.Libdl.var"#dlsym"(), args=("/cache/julia-buildkite-plugin/depots/53de1ea5-6dd3-4008-b976-161c9f34b0ad/artifacts/4db1e58d71ac6bbb35ecc832033264c630d5d3b3/lib/libzstd.so", :ZSTD_versionString), world=0x00000000000097b7)
```
(LinearAlgebra is already broken for `--trim` on master for the same reason)

We do have a plan to fix it, but I don't know when it's going to happen since @keno would like to revamp our ccall semantics (https://github.com/JuliaLang/julia/issues/57931) around the same time

IanButterworth commented:
Ah ok. 
In its current state, along with https://github.com/JuliaSparse/SparseArrays.jl/pull/626 this PR gives
```
julia> @time using Plots
  0.850111 seconds (1.29 M allocations: 74.658 MiB, 6.23% gc time, 2.34% compilation time)
```
compared to results in https://github.com/JuliaLang/julia/pull/58436

IanButterworth commented on stdlib/GMP_jll/test/runtests.jl:
```diff
@@ -3,6 +3,6 @@
 
 @testset "GMP_jll" begin
-    vn = VersionNumber(unsafe_string(unsafe_load(cglobal((:__gmp_version, libgmp), Ptr{Cchar}))))
+    vn = VersionNumber(unsafe_string(unsafe_load(cglobal(dlsym(libgmp, :__gmp_version), Ptr{Cchar}))))
```
Should/could cglobal be taught how to handle a LazyLibrary?

staticfloat replied:
It could, the most annoying thing is just that `cglobal()` is an intrinsic, not a function, so to add an overload for that involves writing C code, not just Julia code.

staticfloat commented on stdlib/GMP_jll/src/GMP_jll.jl:
```diff
@@ -4,49 +4,53 @@
-libgmpxx_handle::Ptr{Cvoid} = C_NULL
 libgmpxx_path::String = ""
 
+_libgmp_dependencies = LazyLibrary[]
+_libgmpxx_dependencies = LazyLibrary[]
```
```suggestion
```

staticfloat commented on stdlib/GMP_jll/src/GMP_jll.jl:
```diff
@@ -4,49 +4,53 @@
 baremodule GMP_jll
 using Base, Libdl
```
```suggestion
using Base, Libdl, CompilerSupportLibraries_jll
```

staticfloat commented on stdlib/GMP_jll/src/GMP_jll.jl:
```diff
@@ -4,49 +4,53 @@
+
+const libgmpxx = LazyLibrary(
+    _libgmpxx_path,
+    dependencies=_libgmpxx_dependencies,
```
```suggestion
    dependencies=[libgmp, libstdcxx, libgcc_s],
```

Turns out `libgmpxx` has a bunch of dependencies that we didn't really capture previously.

staticfloat commented on stdlib/GMP_jll/src/GMP_jll.jl:
```diff
@@ -4,49 +4,53 @@
+)
+
+function eager_mode()
+    dlopen(libgmp)
```
```suggestion
    CompilerSupportLibraries_jll.eager_mode()
    dlopen(libgmp)
```

staticfloat commented on stdlib/libLLVM_jll/src/libLLVM_jll.jl:
```diff
@@ -5,41 +5,41 @@
 
+const libLLVM = LazyLibrary(
+    _libLLVM_path,
+    dependencies=_libLLVM_dependencies,
```
```suggestion
    dependencies=[libstdcxx, libgcc_s, libz],
```

I will note that technically there are all these dependencies, and we should capture them, but it's probably not going to hurt us yet, as `libLLVM` is generally either loaded before Julia code gets to run, or not loaded at all.

staticfloat commented on stdlib/OpenLibm_jll/src/OpenLibm_jll.jl:
```diff
@@ -4,41 +4,41 @@
-libopenlibm_handle::Ptr{Cvoid} = C_NULL
 libopenlibm_path::String = ""
 
+_libopenlibm_dependencies = LazyLibrary[]
```
```suggestion
```

staticfloat commented on stdlib/PCRE2_jll/src/PCRE2_jll.jl:
```diff
@@ -4,41 +4,41 @@
-libpcre2_8_handle::Ptr{Cvoid} = C_NULL
 libpcre2_8_path::String = ""
 
+_libpcre2_8_dependencies = LazyLibrary[]
```
```suggestion
```

staticfloat commented on stdlib/PCRE2_jll/src/PCRE2_jll.jl:
```diff
@@ -4,41 +4,41 @@
 
+const libpcre2_8 = LazyLibrary(
+    _libpcre2_8_path,
+    dependencies=_libpcre2_8_dependencies,
```
```suggestion
```

staticfloat commented on stdlib/OpenLibm_jll/src/OpenLibm_jll.jl:
```diff
@@ -4,41 +4,41 @@
 
+const libopenlibm = LazyLibrary(
+    _libopenlibm_path,
+    dependencies=_libopenlibm_dependencies,
```
```suggestion
```

staticfloat commented on stdlib/SuiteSparse_jll/src/SuiteSparse_jll.jl:
```diff
@@ -4,137 +4,134 @@
 end
 
-function __init__()
+const libamd = LazyLibrary(_libamd_path, dependencies=_libamd_dependencies)
+const libbtf = LazyLibrary(_libbtf_path, dependencies=_libbtf_dependencies)
+const libcamd = LazyLibrary(_libcamd_path, dependencies=_libcamd_dependencies)
+const libccolamd = LazyLibrary(_libccolamd_path, dependencies=_libccolamd_dependencies)
+const libcholmod = LazyLibrary(_libcholmod_path, dependencies=_libcholmod_dependencies)
+const libcolamd = LazyLibrary(_libcolamd_path, dependencies=_libcolamd_dependencies)
+const libklu = LazyLibrary(_libklu_path, dependencies=_libklu_dependencies)
+const libldl = LazyLibrary(_libldl_path, dependencies=_libldl_dependencies)
+const librbio = LazyLibrary(_librbio_path, dependencies=_librbio_dependencies)
+const libspqr = LazyLibrary(_libspqr_path, dependencies=_libspqr_dependencies)
+const libsuitesparseconfig = LazyLibrary(_libsuitesparseconfig_path, dependencies=_libsuitesparseconfig_dependencies)
+const libumfpack = LazyLibrary(_libumfpack_path, dependencies=_libumfpack_dependencies)
```
This is going to be tedious, but you need to express the interdependencies between each of these libraries.  E.g. `libamd` depends on `libsuitesparseconfig`, and `libcholmod` depends on `libamd`, `libcamd`, `libccolamd`, etc...

IanButterworth replied:
I used a couple of bots to produce and check the list. Just pushed. Does it look reasonable?

topolarity commented:
Can / should we add a test that checks the dependency list against the `ldd` (or `DT_NEEDED` entries) of the .so's?

IanButterworth commented:
Would it be enough to check their dlpath?

vtjnash commented:
BinaryBuilder should already audit that, except for a few know broken libraries

topolarity commented:
> BinaryBuilder should already audit that, except for a few know broken libraries

Right, but it would be good to make sure that we haven't fallen out-of-sync with BinaryBuilder

staticfloat commented:
> it would be good to make sure that we haven't fallen out-of-sync with BinaryBuilder

I agree, I added a test that can be run on macOS or Linux.  I (or someone else feeling brave) can make one for Windows as well.  It already catches plenty of minor issues:

```
┌ Warning: Dependency mismatch
│   jll = "SuiteSparse_jll"
│   library = "libbtf"
│   missing_deps = ""
│   extraneous_deps = "libsuitesparseconfig, libamd, libcolamd"
└ @ Main ~/src/julia/test/stdlib_dependencies.jl:169

┌ Warning: Dependency mismatch
│   jll = "SuiteSparse_jll"
│   library = "libldl"
│   missing_deps = ""
│   extraneous_deps = "libsuitesparseconfig"
└ @ Main ~/src/julia/test/stdlib_dependencies.jl:169

┌ Warning: Dependency mismatch
│   jll = "SuiteSparse_jll"
│   library = "libspqr"
│   missing_deps = "libstdc++, libgcc_s"
│   extraneous_deps = "libamd, libcolamd"
└ @ Main ~/src/julia/test/stdlib_dependencies.jl:169

┌ Warning: Dependency mismatch
│   jll = "SuiteSparse_jll"
│   library = "libumfpack"
│   missing_deps = ""
│   extraneous_deps = "libcolamd"
└ @ Main ~/src/julia/test/stdlib_dependencies.jl:169

┌ Warning: Dependency mismatch
│   jll = "MPFR_jll"
│   library = "libmpfr"
│   missing_deps = ""
│   extraneous_deps = "libgmpxx"
└ @ Main ~/src/julia/test/stdlib_dependencies.jl:169
Test Summary:               | Pass  Fail  Total  Time
Stdlib JLL dependency check |   29     5     34  3.9s
RNG of the outermost testset: Random.Xoshiro(0x1329e9815a60ab0e, 0x784c9414423a89d0, 0x1a3aea28fda405f6, 0x8c1dfb6d4dda4f7b, 0xd4081888363c92a9)
ERROR: LoadError: Some tests did not pass: 29 passed, 5 failed, 0 errored, 0 broken.
in expression starting at /home/sabae/src/julia/test/stdlib_dependencies.jl:132
```

Note that there are some slight differences between macOS and Linux, so you'll have to add some conditionals for those things.

IanButterworth commented:
Nice

vtjnash commented:
> Blocked by --trim failure

That should not actually be blocking. The required code dependencies should be vendored into the test directory anyways, rather than relying on changeable state elsewhere, since otherwise we can lose test coverage simply if these package files change, which should not be true of test code.

staticfloat commented on stdlib/LibSSH2_jll/src/LibSSH2_jll.jl:
```diff
@@ -3,48 +3,57 @@
+const libssh2 = LazyLibrary(_libssh2_path, dependencies=_libssh2_dependencies)
+
+function eager_mode()
+    @static if !Sys.iswindows()
```
```suggestion
    Zlib_jll.eager_mode()
    @static if !Sys.iswindows()
```

staticfloat commented on stdlib/SuiteSparse_jll/src/SuiteSparse_jll.jl:
```diff
@@ -2,139 +2,188 @@
+end
+const libumfpack = LazyLibrary(_libumfpack_path; dependencies=_libumfpack_dependencies)
+
+function eager_mode()
```
```suggestion
function eager_mode()
    CompilerSupportLibraries_jll.eager_mode()
```

staticfloat commented:
> The required code dependencies should be vendored into the test directory anyways, rather than relying on changeable state elsewhere, since otherwise we can lose test coverage simply if these package files change, which should not be true of test code.

I don't quite follow what you mean, can you be more explicit?

topolarity commented on stdlib/LibCURL_jll/src/LibCURL_jll.jl:
```diff
@@ -9,41 +9,56 @@ if !(Sys.iswindows() || Sys.isapple())
 libcurl_path::String = ""
 
+if Sys.isfreebsd()
+    _libcurl_dependencies = LazyLibrary[]
```
All of these zero-dependency lists for FreeBSD are surprising - why is it so common for that platform?

IanButterworth replied:
I don't know. They are all empty, actually. 
Also note that the new Zstd_jll also isn't a dep of libLLVM so FreeBSD is erroring.

@staticfloat I didn't see any obvious place the testing could be incorrect?

IanButterworth replied:
Apparently there's a readelf styling difference on freebsd where it shows as `NEEDED` not `(NEEDED)`

IanButterworth replied:
FreeBSD wants libcxxrt for several libs but I don't know what provides that?
```
From worker 9:	┌ Warning: Dependency mismatch
--
  | From worker 9:	│   jll = "GMP_jll"
  | From worker 9:	│   library = "libgmpxx"
  | From worker 9:	│   missing_deps = "libgmp, libc++, libcxxrt, libgcc_s"
  | From worker 9:	│   extraneous_deps = ""
  | From worker 9:	│   actual_deps = "libgmp, libc++, libcxxrt, libgcc_s"
  | From worker 9:	└ @ Main.Test75Main_stdlib_dependencies /usr/home/julia/.buildkite-agent/builds/freebsd13-amdci6-0/julialang/julia-master/julia-

```

IanButterworth replied:
I see it came up here https://github.com/JuliaPackaging/Yggdrasil/pull/2108#issuecomment-731304853 but I don't see how it was resolved

IanButterworth replied:
@giordano pointed out it's an OS lib https://github.com/JuliaPackaging/BinaryBuilder.jl/blob/d10ad161d40260a8aa942b81c100e2bacc462ee9/src/auditor/dynamic_linkage.jl#L215

giordano commented on test/stdlib_dependencies.jl:
```diff
@@ -0,0 +1,196 @@
+
+    # Only keep `(NEEDED)` lines
+    needed_str = Sys.isfreebsd() ? "NEEDED" : "(NEEDED)"
+    libs = filter(l -> occursin(needed_str, l), libs)
```
```suggestion
    libs = filter(contains(needed_str), libs)
```

giordano commented on test/stdlib_dependencies.jl:
```diff
@@ -0,0 +1,197 @@
+
+    # Only keep `(NEEDED)` lines
+    needed_str = Sys.isfreebsd() ? "NEEDED" : "(NEEDED)"
+    libs = filter(occursin(needed_str), libs)
```
That's not what I suggested: https://github.com/JuliaLang/julia/pull/58444#discussion_r2105152619 :wink:

topolarity commented on test/trimming/basic_jll.jl:
```diff
@@ -3,7 +3,7 @@ using Zstd_jll
 function @main(args::Vector{String})::Cint
     println(Core.stdout, "Julia! Hello, world!")
-    fptr = dlsym(Zstd_jll.libzstd_handle, :ZSTD_versionString)
+    fptr = dlsym(Zstd_jll.libzstd, :ZSTD_versionString)
```
That's OK by me - I still consider this a `--trim` regression, but it's worth allowing it so that we can get LazyLibary in for now.

We've got a plan to fix-up LazyLibrary for `--trim` soon enough and at that point we'll bump the test to use the latest standard JLL's (incl. LazyLibrary)

IanButterworth replied:
Ok. Pushed.

IanButterworth replied:
It's the `ccall((:ZSTD_versionString, libzstd()), Cstring, ())` that's segfaulting
https://github.com/JuliaLang/julia/blob/dc4c2dc1f9f28b369ab4d105acabd4de8f95f165/test/trimming/basic_jll.jl#L13

Removing that it outputs
```
Julia! Hello, world!
1.5.7
```

vtjnash replied:
`libzstd()` seems like a typo, since that syntax is known UB

IanButterworth replied:
What's the correct way to load and ccall a lib in the trim world? I'm really not familiar

IanButterworth replied:
I've commented that one out with a fixme.

IanButterworth replied:
Ok. I missed that you meant the whole old Zstd_jll package, not just its lib file.

IanButterworth committed:
copy libzstd over for trimming tests

comment-out bad ccall with a fixme

Revert "comment-out bad ccall with a fixme"

This reverts commit 6fc66a24dd6776991206a26b7fa866d21342be26.

Revert "copy libzstd over for trimming tests"

This reverts commit dc4c2dc1f9f28b369ab4d105acabd4de8f95f165.

vendor the old Zstd_jll

IanButterworth commented:
I haven't figured out why, but load time seems to have regressed on this since https://github.com/JuliaLang/julia/pull/58444#issuecomment-2888030582 (0.85s)

```
julia> @time using Plots
  0.969735 seconds (1.35 M allocations: 76.738 MiB, 5.62% gc time, 1.59% compilation time)
```

Lots of big changes on master since. We need visible key performance metrics like load time on every PR, before merge.

topolarity commented:
Removing the backport label - any `LazyLibrary` changes are not eligible for 1.12 (as decided on the original PR: https://github.com/JuliaLang/julia/pull/57719#issuecomment-2742735196)
